### PR TITLE
Metadata rest server for dashboard

### PR DESCRIPTION
**Pull request by Sami1309**
Changed and added minor functionality to client

### CODE REVIEW

File: dashboard/src/api/resources/Resources.js [hunk 1]
```diff
@@ -31,10 +44,10 @@ export const resourcePaths = Object.freeze({
   "Feature Set": "/feature-sets",
   Model: "/models",
   Transformation: "/transformations",
-  "Training Dataset": "/training-datasets",
+  "Training Dataset": "/training-sets",
   Provider: "/providers",
   User: "/users",
-  "Primary Data": "/primary-data",
+  "Primary Data": "/sources",
```
Comment (simba-git) [inline dashboard/src/api/resources/Resources.js:50]: Why revert this

File: dashboard/src/components/resource-list/ResourceSlice.js [hunk 2]
```diff
@@ -1,49 +1,56 @@
...
-import { resourceTypes } from "api/resources";
+import { resourceTypes, resourceVersions } from "api/resources";
 
 const assertAndCheck = (assertion, errorMessage) => {
   console.assert(assertion, { errorMsg: errorMessage });
   return assertion;
 };
 
-const hasRequiredObjects = (resources) => {
+const hasRequiredObjects = (resources, hasVersions) => {
```
Comment (simba-git) [inline dashboard/src/components/resource-list/ResourceSlice.js:9]: isValidResponse?

File: ff/serving/metadata-rest/metadata-rest-server.go [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+
+func NewMetadataServer(logger *zap.SugaredLogger, client *metadata.Client) (*MetadataServer, error) {
+	logger.Debug("Creating new metadata server")
+	return &MetadataServer{
+		client: client,
+		logger: logger,
+	}, nil
+}
+
+type NameVariant struct {
```
Comment (simba-git) [inline ff/serving/metadata-rest/metadata-rest-server.go:27]: This is unused

File: ff/serving/metadata-rest/metadata-rest-server.go [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+func (m *FetchError) Error() string {
+	return fmt.Sprintf("Error %d: Failed to fetch %s", m.StatusCode, m.Type)
+}
+
+func (m MetadataServer) readFromFeature(feature *metadata.Feature) (map[string]FeatureVariantResource, *FetchError) {
+	variantMap := make(map[string]FeatureVariantResource)
+	variants, err := feature.FetchVariants(m.client, context.Background())
+	if err != nil {
+		fetchError := &FetchError{StatusCode: 500, Type: "feature variants"}
+		m.logger.Errorw("Failed to fetch variants", "Error", err)
```
Comment (simba-git) [inline ff/serving/metadata-rest/metadata-rest-server.go:160]: Use:
fetchError.Error() for the message

File: ff/serving/metadata-rest/metadata-rest-server.go [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+	}
+	return variantMap, nil
+}
+
+func (m MetadataServer) readFromTrainingSet(trainingSet *metadata.TrainingSet) (map[string]TrainingSetVariantResource, *FetchError) {
+	variantMap := make(map[string]TrainingSetVariantResource)
+	variants, err := trainingSet.FetchVariants(m.client, context.Background())
+	if err != nil {
+		fetchError := &FetchError{StatusCode: 500, Type: "training set variants"}
+		m.logger.Errorw("Failed to fetch variants", "Error", err)
```
Comment (simba-git) [inline ff/serving/metadata-rest/metadata-rest-server.go:186]: Use error mesage

File: ff/serving/metadata-rest/metadata-rest-server.go [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+			TrainingSets: variant.TrainingSets(),
+		}
+	}
+	return variantMap, nil
+}
+
+func (m MetadataServer) GetMetadata(c *gin.Context) {
+	switch c.Param("type") {
+	case "features":
+		features, err := m.client.GetFeatures(context.Background(), []string{c.Param("resource")})
```
Comment (simba-git) [inline ff/serving/metadata-rest/metadata-rest-server.go:257]: If client doesn't have a GetFeature( function, I'd add one and use that, makes the code a little cleaner

File: ff/serving/metadata-rest/metadata-rest-server.go [hunk 7]
```diff
@@ -0,0 +1,611 @@
+package main
```
Comment (simba-git) [inline ff/serving/metadata-rest/metadata-rest-server.go:1]: package names shouldnt have a -, go files should use underscore:
I would make this a subpackage of metadata (so a subdirectory of metadata called dashboard)
ff/serving/metadata/dashboard

I'd rename the file to dashboard_metadata.go

File: ff/serving/metadata-rest/metadata-rest-server.go (hunk 4 above)
Comment (Sami1309) [reply ff/serving/metadata-rest/metadata-rest-server.go:160]: Currently the code doesn't print the fetch error here, but rather logs the fetcherror in the method this function returns its error to.

File: ff/serving/metadata-rest/metadata-rest-server.go (hunk 5 above)
Comment (Sami1309) [reply ff/serving/metadata-rest/metadata-rest-server.go:186]: same here

File: ff/serving/metadata-rest/metadata-rest-server.go (hunk 6 above)
Comment (Sami1309) [reply ff/serving/metadata-rest/metadata-rest-server.go:257]: Currently no singular getfeature function or likewise for other resources.

File: ff/serving/metadata-rest/metadata-rest-server.go (hunk 4 above)
Comment (simba-git) [reply ff/serving/metadata-rest/metadata-rest-server.go]: Why not do it like this:
		m.logger.Errorw(fetchError.Error(), "Internal Error", err)
Seems better to me, but if you'd prefer to do it this way, I don't have a super strong opinion

File: ff/serving/metadata-rest/metadata-rest-server.go (hunk 6 above)
Comment (simba-git) [reply ff/serving/metadata-rest/metadata-rest-server.go]: Could you add one, it's essentially the same logic as here

**Review by simba-git — APPROVED** [review]
LGTM, nice work

**Pull request merged**